### PR TITLE
Allow re-request of validation email before token expires

### LINKED ISSUE #6525: Allow user to re-request account validation email before token expires
### Summary

As pointed out by @kschiffer in https://github.com/TheThingsNetwork/lorawan-stack/pull/6524 when a validation for an user's `contact_info`/`email_address` is requested, no other request can be made until the token expires. 

Since the default TTL value of the token is 48 hours it would be in the user's interest for us to support re-sending the same validation a few extra times while the token is still valid. That way if by some reason the initial email does not reach the user, the request of another validation email can be done. 

### Current Situation

We sent a validation request to the email once and can only do it again after the period specified in the `is.user-registration.contact-info-validation.token-ttl`.

### Why do we need this? Who uses it, and when?

It is an edge case but a user might not receive the validation email if:
- The Stack has an internal error while attempting to send the validation email.
- The service used to sent the e-mail fails.

### Proposed Implementation

Cache the validation with a short TTL, use it to assert if we should or not send the same email again. 

If the email (or a hash of the email) is present in the cache then the validation email was sent recently and it might be sent at maximum of X times, if it is not present then it should only be sent after the token is expired.

### Contributing

- [X] I can help by doing more research.
- [X] I can help by implementing the feature after the proposal above is approved.
- [X] I can help by testing the feature before it's released.

### Code of Conduct

- [X] I agree to follow TTN's [Community Code of Conduct](https://github.com/TheThingsNetwork/Manifesto/blob/master/CodeOfConduct.md).

### PR DESCRIPTION
**Pull request by nicholaspcr**
#### Summary

Closes #6525 

#### Changes

- Add `EmailInterval` options to user registration configuration.
- Update request validation method
- Add unit tests for contact_registry
- add test cases for contact_validation_store.
- Update `validations_already_sent` error, now includes `retry_interval`

#### Testing

Manually tested, unit tests.

<details>
  <summary> Test steps </summary>
   
   Run Stack locally with `contact-info-validation.required` as `true` and a `retry-interval` with a low value.

   Config example:
   ```yaml
   is:
     user-registration:
      contact-info-validation:
        required: false
        token-ttl: 48h0m0s
        retry-interval: 0h2m0s
  ```

Perform the following steps:

  1. Create a new non admin user. `ttn-lw-cli usr create <usr_id> --primary-email-address <email> --password <password>`
  2. Request validation `ttn-lw-cli usr contact-info request-validation <usr_id>`
    - Should return an error
  3. Wait until the time set on `retry-interval` has passed.
  4. Request another validation `ttn-lw-cli usr contact-info request-validation <usr_id>`
    

  The last validation request should've worked, as `token-ttl` is the default or a lot larger than the `retry-interval` it means that a new validation email was sent without reseting the validation itself.

  If you want to be even more detailed on the testing you could enable email being written to the local directory and then making sure that there is a new email after requesting a validation. The config example for doing such behavior is listed below, the steps are the same as the ones mentioned before you just need to count the amount of email before and after the validation request.

  ```yaml
  tls:
    source: file
    root-ca: ./ca.pem
    certificate: ./cert.pem
    key: ./key.pem

  is:
    email:
      dir: ".env/emails" # Example of directory path
      provider: "dir"
      sender-address: ""
      sender-name: # Example"employee@thethingsindustries.com"
      templates:
        directory: # Example "<repo_path>/pkg/email/templates"
    user-registration:
      contact-info-validation:
        required: true
        token-ttl: 48h0m0s
        retry-interval: 0h2m0s
  ```
</details>

##### Regressions

This adds the ability to resend the same validation, instead of waiting for the validation to expire in order to send a new one,  I took care to guarantee that the change doesn't break what was the expected behavior.

#### Notes for Reviewers

https://github.com/TheThingsNetwork/lorawan-stack/issues/6567 Is the main reason as of why I did the updates that are related to the contact_info store and registry, its counter intuitive since the field is deprecated but since this is specifically around the behavior of validations it felt appropriate to write these tests. They will change after separating the contact_info from validations. 

#### Checklist

- [ ] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] The steps/process to test this feature are clearly explained including testing for regressions.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [ ] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/ttn/lorawan/v3/identityserver.proto [hunk 1]
```diff
@@ -70,6 +70,8 @@ message IsConfiguration {
 
     message ContactInfoValidation {
       google.protobuf.BoolValue required = 1;
+      google.protobuf.Duration token_ttl = 2;
+      google.protobuf.Duration email_interval = 3;
```
Comment (johanstokking) [inline api/ttn/lorawan/v3/identityserver.proto]: Can you add comments here?

Especially what does `email_interval` mean?

File: pkg/identityserver/contact_info_registry.go [hunk 2]
```diff
@@ -106,17 +107,41 @@ func (is *IdentityServer) requestContactInfoValidation(
...
+		if err != nil {
+			return err
+		}
+
+		validations, err := st.ListValidations(ctx, ids)
+		if err != nil {
+			return err
+		}
+		for _, v := range validations {
+			if v.GetUpdatedAt().AsTime().Before(now.Add(-retryInterval)) {
```
Comment (adriansmares) [inline pkg/identityserver/contact_info_registry.go]: I would make something like `ListRefreshableValidations` to be honest - there is no need to filter here.

File: api/ttn/lorawan/v3/identityserver.proto [hunk 3]
```diff
@@ -70,6 +70,9 @@ message IsConfiguration {
 
     message ContactInfoValidation {
       google.protobuf.BoolValue required = 1;
+      google.protobuf.Duration token_ttl = 2;
+      // RetryInterval is the minimum interval for resending a validation email
```
Comment (adriansmares) [inline api/ttn/lorawan/v3/identityserver.proto]: The comments don't have to be in the same format that we have in Go - in general in the protos we use something like `The minimum interval between validation emails.`

File: pkg/identityserver/bunstore/contact_info_store.go [hunk 4]
```diff
@@ -557,3 +607,35 @@ func (s *contactInfoStore) ExpireValidation(ctx context.Context, pb *ttnpb.Conta
...
+		if errors.IsNotFound(err) {
+			return store.ErrValidationTokenNotFound.WithAttributes("validation_id", pb.Id)
+		}
+		return err
+	}
+
+	_, err = s.DB.NewUpdate().
+		Model(model).
+		WherePK().
+		Column("updated_at").
```
Comment (adriansmares) [inline pkg/identityserver/bunstore/contact_info_store.go]: Is this `Column` needed ? The other two columns are not present here.

File: pkg/identityserver/storetest/contact_info_store.go [hunk 5]
```diff
@@ -144,6 +162,42 @@ func (st *StoreTest) TestContactInfoStoreCRUD(t *T) {
...
+					}
+
+					err = s.RefreshValidation(ctx, original)
+					a.So(err, should.BeNil)
+
+					after, err := s.GetValidation(ctx, validation)
+					if !a.So(err, should.BeNil) || !a.So(after.GetUpdatedAt(), should.NotBeNil) {
+						t.FailNow()
+					}
+
```
Comment (adriansmares) [inline pkg/identityserver/storetest/contact_info_store.go]: Maybe also test that a second validation does not work ?

File: pkg/identityserver/bunstore/contact_info_store.go [hunk 6]
```diff
@@ -534,6 +561,31 @@ func (s *contactInfoStore) GetValidation(
...
+		return nil, err
+	}
+
+	return s.listValidationsBy(ctx, func(q *bun.SelectQuery) *bun.SelectQuery {
+		return q.
+			Where("?TableAlias.entity_id = ?", entityUUID).
+			Where("?TableAlias.entity_type = ?", entityType).
+			Where("?TableAlias.used IS NULL OR ?TableAlias.used = false").
+			Where("?TableAlias.expires_at IS NULL OR ?TableAlias.expires_at > NOW()").
+			Where("?TableAlias.updated_at <= ?", now().Add(-refreshInterval))
```
Comment (adriansmares) [inline pkg/identityserver/bunstore/contact_info_store.go]: I would do the arithmetic using `NOW()` in SQL, not by providing the timestamp here.

**Review by adriansmares — APPROVED** [review]

**Comment by KrishnaIyer** [comment]
@nicholaspcr: Please add testing steps.

**Comment by nicholaspcr** [comment]
Test steps added on the PR's testing section

**Pull request merged**